cloud/scope/powervs_machine.go

```diff
@@ -934,29 +934,45 @@ func (m *PowerVSMachineScope) GetZone() string {
 }
 
 // GetServiceInstanceID returns the service instance id.
-func (m *PowerVSMachineScope) GetServiceInstanceID() string {
+func (m *PowerVSMachineScope) GetServiceInstanceID() (string, error) {
 	if m.IBMPowerVSCluster.Status.ServiceInstance != nil && m.IBMPowerVSCluster.Status.ServiceInstance.ID != nil {
-		return *m.IBMPowerVSCluster.Status.ServiceInstance.ID
+		return *m.IBMPowerVSCluster.Status.ServiceInstance.ID, nil
 	}
 	if m.IBMPowerVSCluster.Spec.ServiceInstanceID != "" {
-		return m.IBMPowerVSCluster.Spec.ServiceInstanceID
+		return m.IBMPowerVSCluster.Spec.ServiceInstanceID, nil
 	}
 	if m.IBMPowerVSCluster.Spec.ServiceInstance != nil && m.IBMPowerVSCluster.Spec.ServiceInstance.ID != nil {
-		return *m.IBMPowerVSCluster.Spec.ServiceInstance.ID
+		return *m.IBMPowerVSCluster.Spec.ServiceInstance.ID, nil
 	}
-	return ""
+	// If we are not able to find service instance id, derive it from name if defined.
+	if m.IBMPowerVSCluster.Spec.ServiceInstance != nil && m.IBMPowerVSCluster.Spec.ServiceInstance.Name == nil {
+		return "", fmt.Errorf("failed to find service instance id as both name and id are not set")
+	}
+	serviceInstance, err := m.ResourceClient.GetServiceInstance("", *m.IBMPowerVSCluster.Spec.ServiceInstance.Name, ptr.To(m.GetZone()))
+	if err != nil {
+		m.Error(err, "failed to get Power VS service instance id", "serviceInstanceName", *m.IBMPowerVSCluster.Spec.ServiceInstance.Name)
+		return "", err
+	}
+	// It's safe to directly dereference GUID as its already done in NewPowerVSMachineScope
+	return *serviceInstance.GUID, nil
```

Comment:
but we have removed this code intentionally before right?

Reply:
@Amulyam24 ^^

Reply:
@mkumatag, this a different function called using PowerVSMachineScope
The function we modified earlier is https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_cluster.go#L388 called using PowerVSClusterScope